### ISSUE TITLE
Adding a removing Z-Wave devices

### ISSUE THREAD

**Issue by WojtekKozaczynski** [BlackLight/platypush#135]
I am using GE Z-Wave smart switch and Z-Wave.Me USB Smart Home Stick, which work fine in Home Assistant (HA). I noticed two issues

1. I attempted to add the switch (again) - I missed that probably because the switch and USB controller have been paired already in HA they both showed up as know devices in platypush.  That put the controller in the synch mode that did not stop. Even when I restarted platypush it went back in the synch mode again. I had to reboot the Pi.

2. When I tried to remove the switch (log enclosed) platypsh did not remove it and went into an unstable mode (from log messages it looks like it was looping). 
[2_platypush.log](https://github.com/BlackLight/platypush/files/4660799/2_platypush.log)

**Comment by BlackLight** [comment]
The issue should now be fixed in [98ced4b4](https://github.com/BlackLight/platypush/commit/98ced4b4e45920af130fa355cfc28830fcff8221) and [1b476151](https://github.com/BlackLight/platypush/commit/1b476151007c187a24812fc7ff712719ab452cac). Let me know if it works for you - if so I'll include them in the next release.

**Comment by WojtekKozaczynski** [comment]
I did a clean install as a user other than ```pi``
```
git clone https://github.com/BlackLight/platypush 
cd platypush
sudo python3 setup.py build install
sudo python3 -m pip install -r requirements.txt
```
I uncommented the Z-Wave module in the requirements file, but I must have missed something (?) because the Web interface is coming up in a reduced format like below

![platypush](https://user-images.githubusercontent.com/18745450/82723253-83d9ec00-9c82-11ea-9c6a-42b48ea31191.JPG)
I can turn the Switch on and off, but I cannot remove the Switch node from the Z-Wave network in that UI format. Here is the log 
[platypush.log](https://github.com/BlackLight/platypush/files/4671265/platypush.log)

BTW when I am a user other than ```pi``` I still I have to run platypush elevated ```sudo platypush``` or it fails with a message that it cannot write to the Z-Wave controller ```/dev/ttyACM0```. So placing the config file in ~/.config/platypush directory still does not work (at least for me).

**Comment by BlackLight** [comment]
It looks like you haven't built the CSS stylesheet. To validate if it's the case:

1. Press F12 in the browser -> Console. You should see an error like `/static/css/dist/webpanel.css not found`.

2. Check if `/usr/local/lib/python3.x/dist-packages/platypush-0.12.7-***/platypush/backend/http/static/css/dist/webpanel.css` exists.

The CSS files should be built by `python3 setup.py build`. Try to run that command alone to see if the files are indeed correctly generated or something is wrong, and make sure that you have `pyScss` installed.

> I can turn the Switch on and off, but I cannot remove the Switch node from the Z-Wave network in that UI format.

I don't see any `zwave.remove*` requests in those logs, although I see the successful requests to `zwave.switch_all`. It could be that the request is not sent from the broken UI. However, keep in mind that the Z-Wave node removal process may not be completely supported. `remove_node_from_group` works, `remove_failed_node` should work if the connection with a node has actually failed, `remove_node` is instead quite shaky - the reason is that it's shaky in the first place in its implementation in `openzwave`. I hope better Python alternatives than `openzwave` come up in the future.

> BTW when I am a user other than pi I still I have to run platypush elevated sudo platypush or it fails with a message that it cannot write to the Z-Wave controller /dev/ttyACM0.

That's because on any Linux system r/w access to devices is regulated through groups. If your user doesn't belong to the device group, then it won't have access to it. On Raspbian the file should look like this:

```bash
ls -l /dev/ttyACM0                                                                                                                                      
crw-rw---- 1 root dialout 166, 0 May 23 15:37 /dev/ttyACM0
```

So add the user `pi` to the `dialout` group and logout/login to get it to work.

**Comment by WojtekKozaczynski** [comment]
From the bottom
- adding the user to the dialout group fixes the problem of writing to /dev/ttyACM0 non-elevated
- still cannot verify, via the Web UI, if/how I can remove a node (see next point). I can try sending an HTTP request to the endpoint or wait until we figure out the next point
- if you prefer I can open a separate issue for this (?) but I cannot get CSS stile sheets in the right directories. In a nutshell
 -- I have pyScss installed 
-- I have included the output of running ```python3 setup.py build``` - no idea what ```Function not found: var:1``` lines mean
-- interestingly when I use ```pip``` to install platypush, the style sheets get properly installed and the Web UI works fine
[python3_setup_dot_py_build.txt](https://github.com/BlackLight/platypush/files/4673003/python3_setup_dot_py_build.txt)

**Comment by BlackLight** [comment]
- If node removals don't work from the UI then testing it from the API can be a good way to see if the problem is with the UI or the backend. Something like:

```shell
curl -XPOST -H 'Content-Type: application/json' -d '{
  "type":"request",
  "action":"zwave.remove_node"
}' http://host:8008/execute
```

Or:

```shell
curl -XPOST -H 'Content-Type: application/json' -d '{
  "type":"request",
  "action":"zwave.remove_failed_node",
  "args": {
    "node_name": "Node name"
  }
}' http://host:8008/execute
```

Keep in mind however that, as mentioned in the link on HA in my previous comment, openzwave still requires you to unpair the Z-Wave device first from the currently configured network through whichever mechanism is provided by the device, before proceeding with the removal from the controller.

- The `Function not found: var:1` warnings are ok. SASS has a bit of a hard time understanding some newer CSS4 constructs (like `--var`) and spits out those warnings, but the CSS code is still correctly generated.

I see that the CSS files for the web panel are correctly generated:

```
copying platypush/backend/http/static/css/dist/webpanel.css -> build/lib/platypush/backend/http/static/css/dist
...
Generating CSS /home/wojtek/platypush/platypush/backend/http/static/css/source/webpanel/index.scss -> /home/wojtek/platypush/platypush/backend/http/static/css/dist/webpanel.css
...
```

But you mentioned that the CSS files aren't present in the platypush installation folder. Just in case, try to remove any previous platypush installation (`sudo pip uninstall platypush`) and manually remove any remaining platypush folder from `/usr/lib/python-3.7/site-packages` and `/usr/local/lib/python-3.7/site-packages`, as it may be that the new installation still tries to get static files from the previous one. Then reinstall the Github version. You can also install the Github version via pip if you wish - `sudo pip install git+https://github.com/BlackLight/platypush.git`.

**Comment by BlackLight** [comment]
Please let me know if the advised solution worked - I'll close the issue in case.

**Comment by WojtekKozaczynski** [comment]
Sorry for the delay in testing the change, but I've been trying to figure out (with stops and starts) the missing style sheets and for that I need to learn how Flask solutions are deployed. In any case, using ```python3 setup.py build``` to build platypush does not put anything in ```/usr/lib/python-3.7/site-packages``` or ```/usr/local/lib/python-3.7/site-packages```. Instead it seems to be putting scripts in ```/usr/local/bin``` and te code in ```/usr/local/lib/python3.7/dist-packages/platypush/platypush-0.12.8-py3.7.egg```.

**Comment by BlackLight** [comment]
Whether the files end up in dist-packages or site-packages depends on the
environment and interpreter configuration. It doesn't matter much where
they are installed, but just make sure that you don't have duplicate
installations because they can cause these types of issues.

BTW you can also install it from pip. I've made a new release yesterday
that also fixes the SQLite errors in the logs.

**Comment by WojtekKozaczynski** [comment]
I have installed platypush on a clean, latest image of Raspbian using pip ```sudo python3 -m pip install 'platypush[zwave,http]'``` and the web UI works fine (the style sheets are installed properly). 
When I removed the GE Z-Wave smart switch by selecting its node in the UI and pressing "-" 
- it properly initiated the controller to wait on input from the switch - the controller was blinking until I pressed synch button on the switch. 
- it then, based on what I see in the log, started timing out and looping. I have included compressed log below with that loop at the end
[platypush.zip](https://github.com/BlackLight/platypush/files/4750315/platypush.zip)

After restarting platypush the smart switch node was gone. Adding the switch works, that is, the controller waits for input and then adds the smart switch node. However, the switch now gets the next consecutive Node Id - removed switch was # 2 and the same switch added again is # 3. I have a log for restart and adding a node, if you want to see it.

**Comment by BlackLight** [comment]
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/platypush/backend/__init__.py", line 236, in run
    self.loop()
  File "/usr/local/lib/python3.7/dist-packages/platypush/backend/zwave/__init__.py", line 259, in loop
    self._process_event(event)
  File "/usr/local/lib/python3.7/dist-packages/platypush/backend/zwave/__init__.py", line 242, in _process_event
    node=ZwavePlugin.node_to_dict(event.args['node']),
  File "/usr/local/lib/python3.7/dist-packages/platypush/plugins/zwave/__init__.py", line 248, in node_to_dict
    } if hasattr(node, 'values') else {},
  File "/usr/local/lib/python3.7/dist-packages/platypush/plugins/zwave/__init__.py", line 247, in <dictcomp>
    for value_id, value in (node.values or {}).items()
  File "/usr/local/lib/python3.7/dist-packages/platypush/plugins/zwave/__init__.py", line 158, in value_to_dict
    'command_class': value.node.get_command_class_as_string(value.command_class),
  File "/usr/local/lib/python3.7/dist-packages/openzwave/node.py", line 366, in get_command_class_as_string
    return self._network.manager.COMMAND_CLASS_DESC[class_id]
KeyError: None
```

This is strange. The command class of a Z-Wave value should not be null, and that breaks the OpenZWave stack. I hope it's just a glitch during the communication - as far as I know the standard doesn't allow it.

I have [pushed a commit](https://github.com/BlackLight/platypush/commit/a18002dea0efb9c54819bb77f7b6fb6033b13983) that should fix this scenario, let me know if it works for you. If `setup.py` for some reason doesn't work you can install the latest git version through pip via `pip install git+https://github.com/BlackLight/platypush.git`.

**Comment by WojtekKozaczynski** [comment]
I uninstalled platypush using pip (sudo python3 -m pip uninstall platypush) since I installed it from PyPI. 

pi@homeAutoOne:/usr/bin $ sudo python3 -m pip uninstall platypush
Uninstalling platypush-0.12.9:
  Would remove:
    /usr/local/bin/platydock
    /usr/local/bin/platypush
    /usr/local/bin/platyvenv
    /usr/local/bin/pusher
    /usr/local/lib/python3.7/dist-packages/platypush-0.12.9.dist-info/*
    /usr/local/lib/python3.7/dist-packages/platypush/*
    /usr/local/lib/python3.7/dist-packages/tests/*
Proceed (y/n)? y
  Successfully uninstalled platypush-0.12.9

I then installed it using pip, but from GitHub (sudo python3 -m pip install git+https://github.com/BlackLight/platypush.git).  This, however, does not install the style sheets again (?)

**Comment by BlackLight** [comment]
This really shouldn't happen. I know however that the logic of the pip
module is quite f*cked up when it comes to git handles. Please try the
following: I've tried multiple times on fresh installations and it must
work.

- Uninstall any previous installations of platypush. Run pip uninstall
platypush multiple times until you get "Skipping platypush as it is not
installed".
- git clone https://github.com/BlackLight/platypush.git
- cd platypush
- [sudo] pip install .

--
Fabio Manganiello <https://www.fabiomanganiello.com>
Software engineer at Adyen.com
PGP key <0x5BCF008FC9D40C68
<http://keyserver.ubuntu.com/pks/lookup?op=get&search=0x5BCF008FC9D40C68>>

**Comment by WojtekKozaczynski** [comment]
I will try the above installation, but in the meantime the following may help find what's going on 
- I deleted everything pip uninstall would 
```
    /usr/local/bin/platydock
    /usr/local/bin/platypush
    /usr/local/bin/platyvenv
    /usr/local/bin/pusher
    /usr/local/lib/python3.7/dist-packages/platypush-0.12.9.dist-info
    /usr/local/lib/python3.7/dist-packages/platypush
    /usr/local/lib/python3.7/dist-packages/tests
```
- I then cloned platypush from GitHub into ```/home/wojtek/platypush``` and ran  ```sudo python3 setup.py build install```. Enclosed is the terminal output of that command. The style sheets are still **not** working, but maybe the output will give you some clues. 
[setup_py_build_install.txt](https://github.com/BlackLight/platypush/files/4762862/setup_py_build_install.txt)

**Comment by WojtekKozaczynski** [comment]
Good and not-so-good news. Deleting all files and directories as above and then running ```sudo python3 -m pip install .``` from the ```home/wojtek/platypush`` directory installs the style sheets and the UI works. The not-so-good news is that removing the smart switch results in a loop again -- see the end of the enclosed log file. 
When I stop and then restart platypush, the smart switch node is gone and I can add it again with the next Node Id # 4.
[platypush.zip](https://github.com/BlackLight/platypush/files/4763018/platypush.zip)

**Comment by BlackLight** [comment]
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/platypush/backend/__init__.py", line 236, in run
    self.loop()
  File "/usr/local/lib/python3.7/dist-packages/platypush/backend/zwave/__init__.py", line 259, in loop
    self._process_event(event)
  File "/usr/local/lib/python3.7/dist-packages/platypush/backend/zwave/__init__.py", line 242, in _process_event
    node=ZwavePlugin.node_to_dict(event.args['node']),
  File "/usr/local/lib/python3.7/dist-packages/platypush/plugins/zwave/__init__.py", line 248, in node_to_dict
    } if hasattr(node, 'values') else {},
  File "/usr/local/lib/python3.7/dist-packages/platypush/plugins/zwave/__init__.py", line 247, in <dictcomp>
    for value_id, value in (node.values or {}).items()
  File "/usr/local/lib/python3.7/dist-packages/openzwave/value.py", line 124, in id_on_network
    separator, self.index)
TypeError: %x format: an integer is required, not NoneType
```

What's happening here is that you are removing a node, but not all of its info is removed at the same time from the OpenZWave stack. It means that when platypush returns the status of the network some information is still there, but the OpenZWave stack breaks because some of the required information to generate `id_on_network` is missing. In your case, looking at [OpenZWave's source code](https://github.com/OpenZWave/python-openzwave/blob/master/src-api/openzwave/value.py#L124), it's either `home_id`, `command_class` or `index`.

I [have pushed a new commit](https://github.com/BlackLight/platypush/commit/fb58c8991182f1343883829f0ca497aa13fbf3df) that should fix this situation, but keep in mind that I'm really driving blindfolded in this case. Your issue is due to a race condition where you have removed a node but some of its info is still accessible, and that shouldn't happen. I can patch the behaviour in Platypush as much as I can to make it more robust (and probably things will work anyway), but this is really an issue on the OpenZWave side:

- They shouldn't use that old-style format string to generate `id_on_network` that can easily break if you feed null values to it.

- The removal behaviour should be atomic and thread-safe - meaning that while they remove something from the Z-Wave database those records should be locked and not accessible by other parts of the application, otherwise we end up with these nasty race conditions.

**Comment by WojtekKozaczynski** [comment]
I suggest that you close the issue since (a) we know how to install Platypush on Pi with style sheets and (b) you have passed the buck on the node removal behavior to OpenZwave :-) I will peek into their project and see if they already know about that issue and if not, I will try to make a link.
